Feature/movie serise details

Noor-Serry opened the pull request:
# Pull Request Template

## Overview
This PR adds the **Movie Details** screen (plus the “casts” screen created by Bilal).

## Changes
- **UI & Navigation**  
  - Created `MovieDetailsScreen`, its `ViewModel`, and `InteractionListener`  
  - Added navigation route for details screen  
- **Domain & Data**  
  - Added `GetMovieDetailsUseCase`  
  - Extended repository & remote data source to fetch full movie details  
  - Created DTOs, entities, and mappers for movie details and videos  

## Video Demo
[▶️ Watch the demo video](https://thechance-workspace.slack.com/files/U08QS6ZNHHD/F095YRETEFQ/screen_recording_20250716_141625.mp4)

## Checklist
- [x] My code follows this project’s style guidelines  
- [x] I’ve tested my changes manually (UI & video playback)  
- [x] This PR introduces a single feature or logical change  

## Additional Notes
Feel free to leave feedback or ask for changes if anything isn’t clear!

Commit:
feature: add movie details screen

This commit introduces the movie details screen, which displays information about a selected movie.

Key changes include:
- Added `MovieDetailsScreen.kt` with UI components for displaying movie details like title, rating, description, cast, and more.
- Created `MovieDetailsViewModel.kt` to handle the logic and state for the movie details screen.
- Implemented navigation to the movie details screen from the search by actor screen.
- Added an `ExpandableText` composable for displaying long descriptions with a "read more" option.
- Updated `NavGraph.kt` and `Route.kt` to include the movie details route.
- Added `MovieDetailsArgs.kt` for passing movie ID to the ViewModel.
- Included necessary UI state classes (`MovieDetailsUiState.kt`, `ActorUiState.kt`, `Extras.kt`, `Selectable.kt`).
- Added `ExtrasSectionItemInfo.kt` to map `Extras` enum to UI display information.
- Updated `viewModelModule.kt` to include `MovieDetailsViewModel` and `MovieDetailsArgs`.
- Modified `Chip.kt` to support centered text alignment.
- Adjusted `ic_play.xml` drawable size and color.
- Changed the `SEARCH_MOVIE_URL` endpoint in `Endpoints.kt`.
- Made a minor adjustment to `AflamiTheme.kt` for handling null activity.

Commit:
feature: add movie details screen

This commit introduces the movie details screen, which displays information about a selected movie.

Key changes include:
- Added `MovieDetailsScreen.kt` with UI components for displaying movie details like title, rating, description, cast, and more.
- Created `MovieDetailsViewModel.kt` to handle the logic and state for the movie details screen.
- Implemented navigation to the movie details screen from the search by actor screen.
- Added an `ExpandableText` composable for displaying long descriptions with a "read more" option.
- Updated `NavGraph.kt` and `Route.kt` to include the movie details route.
- Added `MovieDetailsArgs.kt` for passing movie ID to the ViewModel.
- Included necessary UI state classes (`MovieDetailsUiState.kt`, `ActorUiState.kt`, `Extras.kt`, `Selectable.kt`).
- Added `ExtrasSectionItemInfo.kt` to map `Extras` enum to UI display information.
- Updated `viewModelModule.kt` to include `MovieDetailsViewModel` and `MovieDetailsArgs`.
- Modified `Chip.kt` to support centered text alignment.
- Adjusted `ic_play.xml` drawable size and color.
- Changed the `SEARCH_MOVIE_URL` endpoint in `Endpoints.kt`.
- Made a minor adjustment to `AflamiTheme.kt` for handling null activity.

Noor-Serry committed:
Create README.md

Commit:
Merge remote-tracking branch 'origin/feature/movie_serise_details' into feature/movie_serise_details

# Conflicts:
#	designSystem/src/main/java/com/example/designsystem/components/ExpandableText.kt

Commit:
Merge branch 'develop' of github.com:Amsterdam-Team/Aflami into feature/movie_serise_details

# Conflicts:
#	designSystem/src/main/java/com/example/designsystem/theme/AflamiTheme.kt
#	remoteDatasource/src/main/java/com/example/remotedatasource/client/Endpoints.kt
#	ui/src/main/java/com/example/ui/screens/searchByActor/SearchByActorScreen.kt

Commit:
feature: add Selectable class to viewModel

Commit:
Merge branch 'develop' of github.com:Amsterdam-Team/Aflami into feature/movie_serise_details

Commit:
feature: implement MovieDetails screen

This commit implements the MovieDetails screen with the following features:
- Display movie details including poster, rating, title, categories, release date, movie length, origin country, description, and cast.
- Implement a MovieExtras section with selectable items (More Like This, Reviews, Gallery, Company Production).
- Show corresponding content based on the selected MovieExtra:
    - More Like This: Displays a list of similar movies.
    - Reviews: Shows user reviews for the movie.
    - Gallery: Presents a gallery of movie images.
    - Company Production: Lists production companies involved.
- Allow navigation to MovieDetails screen from SearchByActor screen.
- Rename `imageUrl` to `posterUrl` in `MovieDetailsUiState`.
- Rename `Extras` enum to `MovieExtras`.
- Rename `MovieDetailsIntractionListener` to `MovieDetailsInteractionListener`.

Commit:
Merge branch 'develop' of github.com:Amsterdam-Team/Aflami into feature/movie_serise_details

Belal-Alkhatib committed:
feature: add Actor data class

Belal-Alkhatib committed:
feature: implement movie cast feature

Belal-Alkhatib committed:
Merge branch 'feature/cast-screen' into feature/movie_serise_details

# Conflicts:
#	app/src/main/java/com/amsterdam/aflami/di/viewModelModule.kt
#	ui/src/main/java/com/example/ui/navigation/NavGraph.kt
#	ui/src/main/java/com/example/ui/navigation/Route.kt

Commit:
refactor: Implement movie details functionality

This commit introduces the necessary changes to fetch and display movie details, including reviews and categories.

Key changes:
- Added `GetMovieDetailsUseCase` to orchestrate fetching movie details, reviews, and categories.
- Extended `MovieRepository` and `CategoryRepository` to support fetching movie-specific data.
- Implemented corresponding methods in `RemoteMovieDatasource` and local data sources to fetch movie reviews and categories.
- Added DTOs for movie reviews and related data structures.
- Created `RemoteReviewMapper` to map review DTOs to domain entities.
- Updated `Movie` entity to remove the `categories` field, as categories are now fetched separately.
- Modified `MovieDetailsScreen` to display reviews and apply styling improvements.
- Added `kotlinx-datetime` dependency for date handling.
- Introduced `ActorUiState` and `ProductionCompany` entities (currently unused, likely for future enhancements).

Commit:
Merge branch 'feature/movie_serise_details' of github.com:Amsterdam-Team/Aflami into feature/movie_serise_details

# Conflicts:
#	domain/src/main/java/com/example/domain/repository/MovieRepository.kt
#	entity/src/main/java/com/example/entity/Actor.kt
#	remoteDatasource/src/main/java/com/example/remotedatasource/datasource/RemoteMovieDatasourceImpl.kt
#	repository/src/main/java/com/example/repository/datasource/remote/RemoteMovieDatasource.kt
#	repository/src/main/java/com/example/repository/repository/MovieRepositoryImpl.kt

Commit:
feature: add baseImageUrl to buildConfig

Commit:
feature: add strings for no production company and no reviews

Commit:
fix: Remove text overflow from ExpandableText

Commit:
feature: extend Movie entity

Commit:
feature: implement GetMovieDetailsUseCase and integrate it into DI

Commit:
feature: implement movie details functionality

This commit introduces the following changes:

- Implemented new data sources in `RemoteMovieDatasourceImpl` to fetch similar movies, movie gallery images, production company details, and movie details by ID.
- Added corresponding DTOs (`RemoteMovieGalleryResponse`, `GalleryImageDto`, `ProductionCompanyDto`, `ProductionCompanyResponse`) and mappers (`RemoteGalleryMapper`, `RemoteProductionCompanyMapper`) to handle the new data.
- Updated `MovieRepositoryImpl` to include new methods for fetching movie details, similar movies, gallery, and production companies.
- Modified `RemoteMovieMapper`, `MovieLocalMapper`, `SearchWithMoviesMapper`, and `MovieWithCategoryMapper` to include additional movie details like runtime, origin country, and video availability.
- Adjusted `RemoteCastMapper` and `RemoteReviewMapper` to use `BuildConfig.BASE_IMAGE_URL` for image URLs.
- Updated `LocalMovieDto` to include new movie detail fields.
- Made `NoInternetConnection` composable public in `SearchByCountryScreen`.
- Corrected return type for `getMovieCategories` in `CategoryDao` and updated its implementation in `LocalCategoryDataSourceImpl`.
- Updated DI module to include new mappers.
- Renamed `getCastByMovieId` to `getActorsByMovieId` in `MovieRepository` and its implementation.
- Renamed `getMovieById` to `getMovieDetailsById` in `MovieRepository` and its implementation.

Commit:
feature: implement MovieDetailsScreen
This commit introduces the MovieDetailsScreen, which displays detailed information about a selected movie.

Key changes include:

- **UI Implementation:**
    - Movie details such as poster, title, rating, categories, release date, length, origin country, description, and cast are displayed.
    - Added sections for "More Like This," "Reviews," "Gallery," and "Company Production."
    - Implemented loading and network error states with corresponding UI elements.
    - Added empty states for sections like "Reviews" and "More Like This" when no data is available.
    - Made the "All" button in the cast section clickable to navigate to a dedicated casts screen.
    - Added a placeholder image for movies without a poster.
    - Play button is disabled when no video is available.
    - Implemented navigation back and to the casts screen.
- **ViewModel Integration:**
    - `MovieDetailsViewModel` fetches movie details using `GetMovieDetailsUseCase`.
    - `MovieDetailsUiStateMapper` is used to transform domain models into UI state.
    - Implemented interaction listeners for navigating back, showing all cast, and retrying on network errors.
    - Handled `NoInternetException` to display a network error state.
- **Navigation:**
    - Updated `Route.Cast` to accept a `movieId` for navigation.
- **DI:**
    - Added `MovieDetailsArgs` and `MovieDetailsUiStateMapper` to the Koin module.

Commit:
fix: remove old Room database schema

Belal-Alkhatib committed:
feature: retrieve and display movie cast by ID

Commit:
Merge branch 'develop' of github.com:Amsterdam-Team/Aflami into feature/movie_serise_details

# Conflicts:
#	app/src/main/java/com/amsterdam/aflami/di/repositoryModule.kt
#	app/src/main/java/com/amsterdam/aflami/di/viewModelModule.kt
#	designSystem/src/main/res/values-ar/strings.xml
#	designSystem/src/main/res/values/strings.xml
#	domain/src/main/java/com/example/domain/exceptions/AflamiException.kt
#	domain/src/main/java/com/example/domain/repository/CategoryRepository.kt
#	domain/src/main/java/com/example/domain/repository/MovieRepository.kt
#	entity/src/main/java/com/example/entity/Movie.kt
#	localDatasource/src/main/java/com/example/localdatasource/roomDataBase/datasource/CategoryLocalSourceImpl.kt
#	remoteDatasource/src/main/java/com/example/remotedatasource/datasource/MovieRemoteSourceImpl.kt
#	repository/src/main/java/com/example/repository/datasource/local/LocalCategoryDataSource.kt
#	repository/src/main/java/com/example/repository/datasource/remote/RemoteMovieDatasource.kt
#	repository/src/main/java/com/example/repository/dto/remote/RemoteMovieItemDto.kt
#	repository/src/main/java/com/example/repository/mapper/local/CategoryLocalMapper.kt
#	repository/src/main/java/com/example/repository/mapper/local/MovieLocalMapper.kt
#	repository/src/main/java/com/example/repository/mapper/local/MovieWithCategoryMapper.kt
#	repository/src/main/java/com/example/repository/mapper/local/SearchWithMoviesMapper.kt
#	repository/src/main/java/com/example/repository/mapper/remote/RemoteMovieMapper.kt
#	repository/src/main/java/com/example/repository/repository/MovieRepositoryImpl.kt
#	ui/src/main/res/values-ar/string.xml
#	ui/src/main/res/values/strings.xml

Commit:
refactor: rename RemoteMovieDatasource to MovieRemoteSource

- Renamed the interface `RemoteMovieDatasource` to `MovieRemoteSource` for consistency.
- Updated `MovieRepositoryImpl` to use `MovieRemoteSource`.
- Adjusted related mappers and view models to reflect the interface name change.
- Added `isUnknownError` state to `MovieDetailsUiState` for better error handling.
- Ensured `Movie` entity includes categories.
- Corrected category retrieval in `CategoryRepositoryImpl`.
- Updated `MovieDetailsViewModel` to handle unknown errors.
- Improved error display in `MovieDetailsScreen`.

Commit:
Merge remote-tracking branch 'origin/feature/movie_serise_details' into feature/movie_serise_details

# Conflicts:
#	domain/src/main/java/com/example/domain/useCase/GetMovieCastUseCase.kt

Commit:
Merge branch 'develop' of github.com:Amsterdam-Team/Aflami into feature/movie_serise_details

# Conflicts:
#	entity/src/main/java/com/example/entity/Movie.kt
#	repository/src/main/java/com/example/repository/mapper/local/CategoryLocalMapper.kt
#	repository/src/main/java/com/example/repository/mapper/local/MovieWithCategoryMapper.kt
#	repository/src/main/java/com/example/repository/mapper/remote/MovieRemoteMapper.kt

Commit:
fix: update movie details and search screens with navigation

This commit includes the following changes:

- Updated the `MovieRemoteMapper` to handle cases where genre IDs are not directly available in the remote DTO.
- Added navigation to movie details from the search by country screen.
- Added navigation to movie details from the search by actor screen.
- Updated the `GetMovieDetailsUseCase` to directly use categories from the movie entity.
- Updated the `MovieDetailsViewModel` to remove unknown error handling.
- Updated the `MovieDetailsScreen` to reflect UI state changes and remove redundant error handling.
- Added movie click handling to the search by country, search by actor, and global search view models.
- Updated search screens to include the movie click listener.
- Updated MediaItemUiState to include the movie ID.

Commit:
fix: use movieId argument in getMovieCast

Commit:
Merge branch 'develop' of github.com:Amsterdam-Team/Aflami into feature/movie_serise_details

Noor-Serry committed:
Delete README.md

Commit:
Merge branch 'feature/movie_serise_details' of github.com:Amsterdam-Team/Aflami into feature/movie_serise_details

Commit:
Merge branch 'develop' of github.com:Amsterdam-Team/Aflami into feature/movie_serise_details

# Conflicts:
#	app/src/main/java/com/amsterdam/aflami/di/viewModelModule.kt
#	entity/build.gradle.kts
#	ui/src/main/java/com/example/ui/screens/search/SearchScreen.kt
#	ui/src/main/java/com/example/ui/screens/searchByActor/SearchByActorScreen.kt
#	ui/src/main/java/com/example/ui/screens/searchByCountry/SearchByCountryScreen.kt
#	viewModel/src/main/java/com/example/viewmodel/search/countrySearch/SearchByCountryEffect.kt
#	viewModel/src/main/java/com/example/viewmodel/search/countrySearch/SearchByCountryInteractionListener.kt
#	viewModel/src/main/java/com/example/viewmodel/search/countrySearch/SearchByCountryViewModel.kt
#	viewModel/src/main/java/com/example/viewmodel/search/searchByKeyword/SearchViewModel.kt

nabilBouzineDev commented on repository/src/main/java/com/example/repository/dto/remote/movieGallery/GalleryImageDto.kt:
```diff
@@ -0,0 +1,15 @@
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+
+@Serializable
```
what this dto represents. GalleryImage is something we have in the Ui. Do you have a more readable name for it?

Noor-Serry replied:
we need the image path :
what name do you think is better ?

eslam-99 commented on designSystem/src/main/java/com/example/designsystem/components/ExpandableText.kt:
As Bareq said, this is a component related to the application not the design system

eslam-99 replied:
And need to add preview

ThoriaHamdy99 replied:
And please add a preview for this composable

nabilBouzineDev commented on repository/src/main/java/com/example/repository/mapper/remote/RemoteProductionCompanyMapper.kt:
```diff
@@ -0,0 +1,18 @@
+import com.example.repository.BuildConfig
+import com.example.repository.dto.remote.ProductionCompanyResponse
+
+class RemoteProductionCompanyMapper {
```
For better autocomplete maybe you let the Remote also in post fix like this ProductionCompanyRemoteMapper

nabilBouzineDev commented on viewModel/src/main/java/com/example/viewmodel/movieDetails/MovieDetailsViewModel.kt:
```diff
@@ -0,0 +1,66 @@
+        tryToExecute(
+            action ={
+                getMovieDetailsUseCase.invoke(state.value.movieId)},
+            onSuccess = { movieDetails ->
```
Why not using the reference syntax here like this ::onFetchMovieSuccess

eslam-99 commented on domain/src/main/java/com/example/domain/repository/MovieRepository.kt:
```diff
@@ -1,9 +1,19 @@
     suspend fun getMoviesByActor(actorName: String): List<Movie>
     suspend fun getMoviesByCountryIsoCode(countryIsoCode: String): List<Movie>
+
+    suspend fun getActorsByMovieId(id: Long): List<Actor>
```
why not `movieId ` same as others :)

ThoriaHamdy99 commented on designSystem/src/main/java/com/example/designsystem/components/ExpandableText.kt:
```diff
@@ -0,0 +1,95 @@
+    showMoreStyle: TextStyle = AppTheme.textStyle.label.medium,
+    showMoreColor: Color = AppTheme.color.primary,
+) {
+    var isExpanded by remember { mutableStateOf(false) }
```
pass these attributes as params to be usable component

ThoriaHamdy99 replied:
I mean isExpanded and isClickable

Noor-Serry replied:
this is state is like animation state

eslam-99 commented on entity/src/main/java/com/example/entity/Movie.kt:
I think `MovieDetails` is better to avoid having big entity that will have unused attributes

Noor-Serry replied:
I did not got you ?

eslam-99 replied:
I mean to have another entity for `MovieDetails`, or what do you think?
- I think its out put class

eslam-99 commented on remoteDatasource/src/main/java/com/example/remotedatasource/datasource/MovieRemoteSourceImpl.kt:
```diff
@@ -60,6 +65,55 @@ class MovieRemoteSourceImpl(
+    override suspend fun getMovieReviews(movieId: Long): ReviewsResponse {
+        return safeCall<ReviewsResponse> {
+            val response = ktorClient.get("movie/$movieId/reviews") {
+            }
```
Why empty curly brackets everywhere?

ThoriaHamdy99 commented on designSystem/src/main/res/drawable/ic_play.xml:
```diff
@@ -1,9 +1,9 @@
-      android:pathData="M13.5,6.402C15.5,7.557 15.5,10.443 13.5,11.598L4.5,16.794C2.5,17.949 0,16.506 0,14.196V3.804C0,1.494 2.5,0.051 4.5,1.206L13.5,6.402Z"
-      android:fillColor="#000000"/>
+      android:pathData="M22.5,9.67C25.833,11.594 25.833,16.406 22.5,18.33L7.5,26.99C4.167,28.915 0,26.509 0,22.66V5.34C0,1.491 4.167,-0.915 7.5,1.01L22.5,9.67Z"
+      android:fillColor="#D85895"/>
```
please make the fill color to be black as default to handle the color of the icon from the UI

Noor-Serry replied:
what is the different ? can you explain ?

eslam-99 commented on remoteDatasource/src/main/java/com/example/remotedatasource/datasource/MovieRemoteSourceImpl.kt:
```diff
@@ -60,6 +65,55 @@ class MovieRemoteSourceImpl(
+    }
+
     private companion object {
         const val SEARCH_MOVIE_URL = "search/movie"
```
As I remember, there is a file for endpoints

Noor-Serry replied:
I did not find this file

ThoriaHamdy99 commented on domain/src/main/java/com/example/domain/useCase/GetMovieCastUseCase.kt:
```diff
@@ -0,0 +1,16 @@
+class GetMovieCastUseCase(private val movieRepository: MovieRepository) {
+
+    suspend operator fun invoke(movieId: Long?): List<Actor>{
+        if (movieId == null) throw NoCastFoundException()
```
we should filter the nullable movies from the data layer and no nullable movies will be passed to the domain and the Ui will not have nullable movies at all

ThoriaHamdy99 replied:
So, it will be not this kind of exception

ThoriaHamdy99 commented on domain/src/main/java/com/example/domain/useCase/GetMovieCastUseCase.kt:
```diff
@@ -0,0 +1,16 @@
+
+        return movieRepository.getActorsByMovieId(movieId)
+            .sortedByDescending { it.popularity }
+            .ifEmpty { throw NoCastFoundException() }
```
As Bareq mentioned the empty is not an exception you can return an empty list and the UI will check on it

ThoriaHamdy99 commented on domain/src/main/java/com/example/domain/useCase/GetMovieDetailsUseCase.kt:
```diff
@@ -0,0 +1,40 @@
+class GetMovieDetailsUseCase(
+    private val movieRepository: MovieRepository,
+) {
+    suspend operator fun invoke(movieId: Long): MovieDetails {
```
Are you asked on the UI/UX about the loading, it will be on all screen or not ?

Noor-Serry replied:
there are one shape for loading across the whole app

ThoriaHamdy99 commented on entity/src/main/java/com/example/entity/Movie.kt:
```diff
@@ -13,4 +13,7 @@ data class Movie(
     override val rating: Float,
     override val popularity: Double,
+    val originCountry: String,
+    val movieLength: Int,
```
I didn't get it. MovieLength? what do you mean by this

Noor-Serry replied:
like 110 minutes

tasneem-hakeem replied:
I think you can name it duration or runtime

ThoriaHamdy99 commented on entity/src/main/java/com/example/entity/ProductionCompany.kt:
```diff
@@ -0,0 +1,8 @@
+
+data class ProductionCompany(
+    val id: Long,
+    val image: String,
```
make it imageUrl please as its more readable

ThoriaHamdy99 commented on entity/src/main/java/com/example/entity/Review.kt:
```diff
@@ -0,0 +1,13 @@
+data class Review(
+    val id : Long,
+    val author: String,
+    val username: String,
```
username of what?

Noor-Serry replied:
what the post include

ThoriaHamdy99 commented on entity/src/main/java/com/example/entity/Review.kt:
```diff
@@ -0,0 +1,13 @@
+    val rating: Float,
+    val content: String,
+    val date: LocalDate,
+    val imageUrl: String?,
```
it can be empty not null? nullable will be very tricky here and will reflect for a lot nullable checks

ThoriaHamdy99 commented on localDatasource/src/main/java/com/example/localdatasource/roomDataBase/daos/CategoryDao.kt:
```diff
@@ -20,4 +20,6 @@ interface CategoryDao {
 
     @Query("SELECT * FROM ${DatabaseContract.TV_SHOW_CATEGORY_TABLE}")
     suspend fun getAllTvShowCategories(): List<LocalTvShowCategoryDto>
+    @Query("SELECT * FROM movie_categories WHERE categoryId = :movieId")
```
Why categoryId = :movieId --> what do you mean it's a diff Ids

ThoriaHamdy99 commented on localDatasource/src/main/java/com/example/localdatasource/roomDataBase/datasource/CategoryLocalSourceImpl.kt:
```diff
@@ -23,6 +23,10 @@ class CategoryLocalSourceImpl(
     override suspend fun getTvShowCategories(): List<LocalTvShowCategoryDto> {
         return dao.getAllTvShowCategories()
     }
+    override suspend fun getCategoriesByMovieId(movieId: Long): List<LocalMovieCategoryDto> {
```
we can rename it to be getCategoryRelatedToMovie() ?

ThoriaHamdy99 commented on remoteDatasource/src/main/java/com/example/remotedatasource/datasource/MovieRemoteSourceImpl.kt:
```diff
@@ -60,6 +65,55 @@ class MovieRemoteSourceImpl(
         }
     }
 
+    override suspend fun getCastByMovieId(id: Long): RemoteCastAndCrewResponse {
```
the param id can be movieId for redability

ThoriaHamdy99 replied:
is RemoteCastAndCrewResponse a relation?

eslam-99 commented on repository/src/main/java/com/example/repository/datasource/remote/MovieRemoteSource.kt:
```diff
@@ -1,9 +1,31 @@
+
+    suspend fun getMoviesByCountryIsoCode(
+        countryIsoCode: String
+    ): RemoteMovieResponse
```
A 1-line function declaration is enough here

eslam-99 commented on repository/src/main/java/com/example/repository/datasource/remote/MovieRemoteSource.kt:
```diff
@@ -1,9 +1,31 @@
+        countryIsoCode: String
+    ): RemoteMovieResponse
+
+    suspend fun getCastByMovieId(id: Long): RemoteCastAndCrewResponse
```
Also `movieId` here

eslam-99 commented on repository/src/main/java/com/example/repository/dto/remote/GenreDto.kt:
```diff
@@ -0,0 +1,9 @@
+
+@Serializable
+data class GenreDto(
+    val id: Int = 0,
```
Need to add `@SerialName`

eslam-99 commented on repository/src/main/java/com/example/repository/dto/remote/ProductionCompanyDto.kt:
```diff
@@ -0,0 +1,14 @@
+
+@Serializable
+data class ProductionCompanyDto(
+    val id: Long,
```
Ensure to add `@SerialName` to all attributes

eslam-99 commented on repository/src/main/java/com/example/repository/dto/remote/movieGallery/GalleryImageDto.kt:
```diff
@@ -0,0 +1,15 @@
+@Serializable
+data class GalleryImageDto(
+    @SerialName("aspect_ratio") val aspectRatio: Double,
+    val height: Int,
```
Also `@SerialName`

eslam-99 commented on repository/src/main/java/com/example/repository/dto/remote/movieGallery/RemoteMovieGalleryResponse.kt:
```diff
@@ -0,0 +1,11 @@
+
+@Serializable
+data class RemoteMovieGalleryResponse(
+    val id: Long,
```
Enough `@SerialName` 😂

ThoriaHamdy99 commented on repository/src/main/java/com/example/repository/datasource/local/CategoryLocalSource.kt:
```diff
@@ -6,7 +6,7 @@ import com.example.repository.dto.local.LocalTvShowCategoryDto
-
     suspend fun upsertTvShowCategories(categories: List<LocalTvShowCategoryDto>)
     suspend fun getTvShowCategories(): List<LocalTvShowCategoryDto>
+    suspend fun getCategoriesByMovieId(movieId : Long) : List<LocalMovieCategoryDto>
```
Also, this function needs to be more readable

eslam-99 commented on repository/src/main/java/com/example/repository/mapper/local/CategoryLocalMapper.kt:
```diff
@@ -33,6 +34,24 @@ class CategoryLocalMapper {
         return localMovieCategory.categoryId.mapToMovieCategory()
     }
 
+    fun mapToLocalTvShowCategory(remoteTvShowCategoryDto: RemoteCategoryDto): LocalTvShowCategoryDto {
```
We need to delete `map` keyword from any mapper

Commit:
refactor: improve code quality and update UI elements

This commit includes several refactoring changes to enhance code quality and consistency:

- Renamed `image` to `imageUrl` in `Category`, `ProductionCompany`, and `Review` entities for clarity.
- Updated mappers and repositories to reflect the `imageUrl` change.
- Removed unused `getMovieCategories(movieId)` method from `CategoryRepository` and its implementations.
- Simplified `GetMovieCastUseCase` by removing unnecessary null check.
- Renamed remote mappers for consistency (e.g., `RemoteCastMapper` to `CastRemoteMapper`).
- Replaced custom loading and no internet connection components with centralized `LoadingContainer` and `NoNetworkContainer` in various screens (`MovieDetailsScreen`, `SearchByActorScreen`, `CastScreen`).
- Added previews for `ExpandableText` component.
- Introduced `safeCall` utility function in `remoteDatasource` for handling API calls.
- Minor UI updates in `SearchScreen` and `SearchViewModel`.
- Removed unused `GlobalSearchViewModel` from `viewModelModule`.
- Updated test cases to align with entity changes.

Commit:
refactor: improve code quality and update UI elements

This commit includes several refactoring changes to enhance code quality and consistency:

- Renamed `image` to `imageUrl` in `Category`, `ProductionCompany`, and `Review` entities for clarity.
- Updated mappers and repositories to reflect the `imageUrl` change.
- Removed unused `getMovieCategories(movieId)` method from `CategoryRepository` and its implementations.
- Simplified `GetMovieCastUseCase` by removing unnecessary null check.
- Renamed remote mappers for consistency (e.g., `RemoteCastMapper` to `CastRemoteMapper`).
- Replaced custom loading and no internet connection components with centralized `LoadingContainer` and `NoNetworkContainer` in various screens (`MovieDetailsScreen`, `SearchByActorScreen`, `CastScreen`).
- Added previews for `ExpandableText` component.
- Introduced `safeCall` utility function in `remoteDatasource` for handling API calls.
- Minor UI updates in `SearchScreen` and `SearchViewModel`.
- Removed unused `GlobalSearchViewModel` from `viewModelModule`.
- Updated test cases to align with entity changes.

Commit:
Merge branch 'develop' of github.com:Amsterdam-Team/Aflami into feature/movie_serise_details

# Conflicts:
#	app/src/main/java/com/amsterdam/aflami/di/repositoryModule.kt
#	app/src/main/java/com/amsterdam/aflami/di/useCaceModule.kt
#	app/src/main/java/com/amsterdam/aflami/di/viewModelModule.kt
#	repository/src/main/java/com/example/repository/repository/MovieRepositoryImpl.kt

Commit:
Refactor: Simplify DI modules and improve code clarity

This commit refactors the dependency injection modules (`viewModelModule`, `useCaseModule`, `repositoryModule`, `localDataSourceModule`) by leveraging Koin's `singleOf` and `bind` for more concise and readable definitions.

Additionally, the following changes were made:

-   Renamed `movieId` to `selectedMovieId` in `SearchUiState` and `SearchViewModel` for better clarity.
-   Added `onMovieClicked` callback to `MoviesVerticalGrid` to handle movie click events in the "Search by Country" screen.
-   Enabled navigation to movie details from the "Search by Country" screen.
-   Removed unused `dispatcher` from `MovieRepositoryImpl`.
-   Improved parameter naming in `MovieRepository` interface (`id` to `movieId`).

Commit:
feat: add SerialName annotations for kotlinx serialization

This commit adds `@SerialName` annotations to DTO classes used with kotlinx.serialization. This ensures correct mapping between JSON field names and Kotlin property names, especially when they differ.

Commit:
fix: add SerialName annotations to GalleryImageDto

Commit:
fix: rename id to movieId in getCastByMovieId

The parameter name `id` was changed to `movieId` in the `getCastByMovieId` function in both `MovieRemoteSource.kt` and `MovieRemoteSourceImpl.kt` for better clarity.

Commit:
style: reformat code in MovieRemoteSource

Commit:
style: remove unnecessary space in GenreDto and ProductionCompanyDto

eslam-99 commented on domain/src/main/java/com/example/domain/exceptions/AflamiException.kt:
```diff
@@ -17,4 +17,5 @@ class NoSuggestedCountriesException : AflamiException()
 class NoSearchByKeywordResultFoundException : AflamiException()
 class NoSearchByActorResultFoundException : AflamiException()
 
+class NoCastFoundException: AflamiException()
```
Bareq said it doesn't make sense to have exception on empty state and we will remove states like this, just return empty list

eslam-99 commented on repository/src/main/java/com/example/repository/repository/MovieRepositoryImpl.kt:
```diff
@@ -142,4 +153,27 @@ class MovieRepositoryImpl(
         )
     }
 
+    override suspend fun getActorsByMovieId(id: Long): List<Actor> {
```
No consistency between compact and normal functions, and also replace id with movieId

eslam-99 commented on ui/src/main/java/com/example/ui/screens/movieDetails/ExtrasSectionItemInfo.kt:
```diff
@@ -0,0 +1,20 @@
+import com.example.viewmodel.movieDetails.MovieExtras
+
+data class ExtrasSectionItemInfo(
+    @DrawableRes val iconResId: Int,
```
Mentors & Bareq recommend to use painter

Noor-Serry replied:
its not the case here

eslam-99 replied:
I know, Bareq commented on a button that doesn't take anything except icon, he still wants painter

Noor-Serry replied:
its data class not composable function its different case also I asked Tareq about this solution in the meeting and and he said its okay .

eslam-99 commented on ui/src/main/java/com/example/ui/screens/movieDetails/MovieDetailsScreen.kt:
```diff
@@ -0,0 +1,693 @@
+                        .background(AppTheme.color.surface)
+                ) {
+
+                    PlayButton(
```
I guess you didn't notice the indicator and the padding around play button
<img width="879" height="767" alt="image" src="https://github.com/user-attachments/assets/e09084f5-ff6a-4424-8604-e77a0012b755" />

Noor-Serry replied:
the indictor its some time exists and some time disappear I did not should I put it or not and what is the benfits of it 
<img width="531" height="164" alt="image" src="https://github.com/user-attachments/assets/25f08946-a124-4229-badf-1b7e6e630ffc" />

eslam-99 replied:
If the movie has multiple images so this considered as a slider and this its indicator, right?

Noor-Serry replied:
its not wrote in the design and if it that what should we do then we should have list of posters Url in the domain

eslam-99 replied:
Okay if API returns a single poster, just see the padding around play button

Noor-Serry replied:
I increase it and this padding is not specified in the design

eslam-99 commented on ui/src/main/java/com/example/ui/screens/movieDetails/MovieDetailsScreen.kt:
```diff
@@ -0,0 +1,693 @@
+                            .padding(horizontal = 16.dp)
+                            .offset(y = (-20).dp)
+                    ) {
+                        Text(
```
Putting all these components inside a single `item` in `LazyColumn` is a waste for `LazyColumn` benifits

Noor-Serry replied:
I did not got you what is the problem with it ?

Noor-Serry replied:
can you explain more ?

eslam-99 replied:
You already know what `LazyColumn` and how it works, I just saw `item` inside it but the item wasn't just an item it was a column of items, why not extracting all these items to the `LazyColumn`

Noor-Serry replied:
I just wrap them with lazy column because we can not make two lazy column

eslam-99 replied:
I mean to just move them up to the parent `LazyColumn`

eslam-99 commented on ui/src/main/java/com/example/ui/screens/movieDetails/MovieDetailsScreen.kt:
```diff
@@ -0,0 +1,693 @@
+    }
+}
+
+fun LazyListScope.CompanyProductionSection(companies: List<ProductionCompanyUiState>) {
```
Separate these components into other new files as this screen is huge, also try to use `internal` access modifier as possible as we use modularization

eslam-99 commented on ui/src/main/java/com/example/ui/screens/movieDetails/MovieDetailsScreen.kt:
```diff
@@ -0,0 +1,693 @@
+                    drawContent()
+                    drawRect(
+                        brush = Brush.verticalGradient(
+                            colors = overlayDarkColor
```
You can use color directly, why incapsulating into a variable for single use

Noor-Serry replied:
I can not call composable thing from non composable scope

eslam-99 commented on ui/src/main/java/com/example/ui/screens/movieDetails/MovieDetailsScreen.kt:
```diff
@@ -0,0 +1,693 @@
+                val strokeWidth = 1.dp.toPx()
+                val y = size.height - strokeWidth / 2
+                drawLine(
+                    color = strokeColor,
```
Same here

Noor-Serry replied:
I can not call composable thing from non composable scope

eslam-99 commented on viewModel/src/main/java/com/example/viewmodel/cast/CastInteractionListener.kt:
```diff
@@ -0,0 +1,6 @@
+
+interface CastInteractionListener {
+    fun onNavigateBackClicked()
+    fun onRetryQuestClicked()
```
Request not Quest

eslam-99 commented on viewModel/src/main/java/com/example/viewmodel/common/Selectable.kt:
```diff
@@ -0,0 +1,6 @@
+package com.example.viewmodel.common
```
Bareq mentioned to use `shared` instead of `common`

eslam-99 commented on viewModel/src/main/java/com/example/viewmodel/movieDetails/MovieDetailsIntractionListener.kt:
```diff
@@ -0,0 +1,8 @@
+package com.example.viewmodel.movieDetails
+
+interface MovieDetailsInteractionListener  {
+    fun onMovieExtrasClicked(movieExtras: MovieExtras)
```
Bareq said its name as `onClickMovieExtras` is better as (on + verb + something)

Noor-Serry replied:
I think onMovieExtrasClicke also as bareq mention when I want the mapper or data source I will write movieRemoteDataSource 
so when write movieI want to see remote or local not writing it with this pattern 
remote so I will see the whole remote mapper or data source 
I hope you got me in this point

Noor-Serry replied:
<img width="753" height="297" alt="image" src="https://github.com/user-attachments/assets/78697aba-0f42-4df7-824b-4a904a73b13f" />
that what I talk about it 
if I write the on + verb it will show me a lot of options 
but on + something + the action its better

eslam-99 replied:
I know the case of Datasources and I agree with it, but Bareq commented specifically on `InteractionListener` like this

Noor-Serry replied:
Is there a mistake in what I said?

eslam-99 replied:
AndroidStudio not sticked to get exact word you type, for example, I have `onShowMovies` & `onShowTvShows`, if I typed `onMo` it still suggesting `onShowMovies` to you
I suggest posting on squad channel

eslam-99 commented on viewModel/src/main/java/com/example/viewmodel/cast/CastInteractionListener.kt:
```diff
@@ -0,0 +1,6 @@
+package com.example.viewmodel.cast
+
+interface CastInteractionListener {
+    fun onNavigateBackClicked()
```
Also here

Noor-Serry replied:
same as upove

eslam-99 commented on viewModel/src/main/java/com/example/viewmodel/movieDetails/MovieDetailsUiState.kt:
```diff
@@ -0,0 +1,65 @@
+)
+
+
+enum class MovieExtras {
```
Bareq suggested to move all these classes and enums inside `MovieDetailsUiState` to avoid miss-use of a state of other UI states from other similar screens

eslam-99 commented on viewModel/src/main/java/com/example/viewmodel/movieDetails/MovieDetailsViewModel.kt:
```diff
@@ -0,0 +1,69 @@
+        }
+    }
+
+    override fun onShowAllCastClicked() {
```
Some functions are compact and some are normal

Noor-Serry replied:
what is the problem if its one line then its okay to be compact  function if it more then one line then its normal function

Commit:
refactor: rename interaction listener methods

Rename interaction listener methods in CastViewModel, CastInteractionListener, MovieDetailsViewModel, and MovieDetailsInteractionListener for clarity.

Update MovieDetailsScreen and CastScreen to use the new method names.

Extract UI components in MovieDetailsScreen into separate files within the `com.example.ui.screens.movieDetails.components` package.

Commit:
refactor: move UI state classes to shared package

Moved `MediaItemUiState` and `Selectable` from the `common` package to a new `shared` package within the `viewModel` module.

This change aims to better organize shared UI state components that are used across multiple ViewModels and UI screens. Updated relevant import statements in `SearchUiState`, `SearchViewModel`, `MovieDetailsUiState`, `MovieDetailsUiStateMapper`, `SearchScreen`, and `MovieExtrasSection`.

Commit:
refactor: nest helper classes in MovieDetailsUiState

This commit refactors the `MovieDetailsUiState` by nesting related helper classes (like `ActorUiState`, `MovieExtras`, etc.) within it.

This improves code organization and readability by grouping related UI state components together.

Commit:
refactor: remove NoCastFound state from CastScreen

The `CastScreen` will now display a generic "No Data" message if the cast list is empty, instead of a specific "No Cast Found" error.

This change simplifies the error handling logic in `CastViewModel` and removes the need for a separate `NoCastFoundException`.

The `CastItemUiState` and `CastErrorUiState` have been nested within `CastUiState` for better organization.

Commit:
Delete unnecessary blank line in MovieDetailsUiState

Commit:
rename 'id' to 'movieId' in getActorsByMovieId

The parameter 'id' in the `getActorsByMovieId` function within `MovieRepositoryImpl` has been renamed to `movieId` for clarity and consistency.

tasneem-hakeem commented on entity/src/main/java/com/example/entity/Review.kt:
```diff
@@ -0,0 +1,13 @@
+
+data class Review(
+    val id : Long,
+    val author: String,
```
Isn't reviewerName more accurate here?

tasneem-hakeem replied:
The same for username to be reviewerUsername?

Commit:
rename entity properties

Rename `movieLength` to `runTime` in `Movie.kt`.
Rename `author` to `reviewerName` and `username` to `reviewerUsername` in `Review.kt`.

Commit:
refactor: rename MovieDetailsInteractionListener method names

This commit refactors the method names in the MovieDetailsInteractionListener interface and its implementations. The changes aim to improve clarity and consistency by:

- Renaming `onMovieExtrasClicked` to `onClickMovieExtras`.
- Renaming `onShowAllCastClick` to `onClickShowAllCast`.
- Renaming `onBackClick` to `onClickBack`.
- Renaming `onRetryQuestClick` to `onClickRetryRequest`.

Additionally, this commit renames:
- `movieLength` to `runTime` in `MovieRemoteMapper` and related mappers (`MovieLocalMapper`, `SearchWithMoviesMapper`, and `MovieDetailsUiStateMapper`).
- `author` to `reviewerName` and `username` to `reviewerUsername` in `ReviewRemoteMapper` and `MovieDetailsUiStateMapper`.

Commit:
refactor: rename cast screen interaction methods

Rename `onNavigateBackClicked` to `onClickNavigateBack` and `onRetrySearchClick` to `onClickRetrySearch` in the `CastInteractionListener`, `CastViewModel`, and `CastScreen` for consistency.

eslam-99 commented on repository/src/main/java/com/example/repository/repository/MovieRepositoryImpl.kt:
```diff
@@ -142,4 +153,27 @@ class MovieRepositoryImpl(
         )
     }
 
+    override suspend fun getActorsByMovieId(movieId: Long): List<Actor> {
```
Need consistency for compact and normal functions

Commit:
refactor: simplify getActorsByMovieId function in MovieRepositoryImpl

tasneem-hakeem commented on viewModel/src/main/java/com/example/viewmodel/movieDetails/MovieDetailsUiState.kt:
```diff
@@ -0,0 +1,65 @@
+        val content: String = "",
+        val date: String = "",
+        val imageUrl: String? = "",
+    )
```
These UI states like ActorUiState, ReviewUiState, SimilarMovieUiState, and ProductionCompanyUiState are also used in the series details screen.
Wouldn't it make more sense to move them outside MovieDetailsUiState into a shared or common package so both movie and series screens can use them without duplication?

Noor-Serry replied:
any class you need from those classes you can extract it in different package with name shared

Commit:
refactor: make sections private in MovieDetailsScreen

The following sections are now private:
- ReviewSection
- MoreLikeSection
- CompanyProductionSection
- CastSection

Commit:
refactor: extract movie details UI states to shared package

The following UI state classes have been moved to a shared package for potential reuse in series details:

- `ActorUiState`
- `ProductionCompanyUiState`
- `ReviewUiState`
- `SimilarMovieUiState`

Additionally, the following composables in `MovieDetailsScreen` have been extracted into their own files within the `components` package:

- `ReviewSection`
- `MoreLikeSection`
- `CompanyProductionSection`
- `CastSection`
